### PR TITLE
V3 Upgrade Handling

### PR DESCRIPTION
**Pull request by fragwuerdig**
## Summary of changes

Apply v3 upgrade mechanics. Following items are needed in this branch:

- [x] create v3 upgrade handling
- [ ] create ibc-go v4 upgrade handler
- [ ] create uprade handler for min_initial_deposit

### CODE REVIEW

File: app/app.go [hunk 1]
```diff
@@ -399,7 +399,7 @@ func GetMaccPerms() map[string][]string {
 
 func (app *TerraApp) setupUpgradeHandlers() {
 	app.UpgradeKeeper.SetUpgradeHandler(
-		v2.UpgradeName,
-		v2.CreateV2UpgradeHandler(app.mm, app.configurator),
+		v3.UpgradeName,
```
Comment (nghuyenthevinh2000) [inline app/app.go:402]: you just need to add new

```
app.UpgradeKeeper.SetUpgradeHandler(
		v3.UpgradeName,
                 v3.CreateV3UpgradeHandler(app.mm, app.configurator),
)
```

and keep v2 intact

Comment (nghuyenthevinh2000) [reply app/app.go:402]: https://github.com/osmosis-labs/osmosis/blob/main/app/app.go#L447

like here, osmosis use a loop to add all upgrade version handler

**Review by nghuyenthevinh2000 — CHANGES_REQUESTED** [review]

**Comment by fragwuerdig** [comment]
hmm .. upgrade test breaks as it seems. Is it the test or the upgrade? @nghuyenthevinh2000 ?

File: app/app.go (hunk 1 above)
Comment (ZaradarBH) [reply app/app.go:402]: I like the Osmosis approach

Comment (fragwuerdig) [reply app/app.go]: Implemented the requested code changes.

File: app/app.go [hunk 2]
```diff
@@ -58,8 +62,13 @@ import (
...
 const appName = "TerraApp"
 
-// DefaultNodeHome defines default home directories for terrad
-var DefaultNodeHome string
+var (
+	// DefaultNodeHome defines default home directories for terrad
+	DefaultNodeHome string
+
+	// Upgrades defines upgrades to be applied to the network
+	Upgrades = []upgrades.Upgrade{v2.Upgrade, v3.Upgrade}
```
Comment (ZaradarBH) [inline app/app.go:70]: You might want to turn this into a constant, like the appName above

**Review by ZaradarBH — APPROVED** [review]
LGTM. Just a single suggestion :)

**Comment by ZaradarBH** [comment]
@fragwuerdig we merged the min deposit PR and approved this. You will however need to re-integrate main to align changes between the two PR branches. Once that is done we can safely merge this PR :flex:

File: app/upgrades/types.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/cosmos/cosmos-sdk/types/module"
+	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/classic-terra/core/app/keepers"
+)
+
+// BaseAppParamManager defines an interrace that BaseApp is expected to fullfil
+// that allows upgrade handlers to modify BaseApp parameters.
+type BaseAppParamManager interface {
```
[the remark this answers is not in the document]
Comment (fragwuerdig) [reply app/upgrades/types.go:15]: uhm... yeah. So `BaseApp` fulfills the interface. And I thought that maybe in future upgrades we wanna have access to features of `BaseApp`? IDK, maybe we don't. Tell me :D

[the remark this answers is not in the document]
Comment (ZaradarBH) [reply app/upgrades/types.go:15]: Thats what I assumed you was  trying to do. We just wanted to verify the intent. I personally think its a good idea :)

**Review by nghuyenthevinh2000 — APPROVED** [review]

**Comment by fragwuerdig** [comment]
@nghuyenthevinh2000: Please, help! I can run the upgrade test locally on my computer by invoking:

`make clean upgrade-start-localnet && ./contrib/updates/upgrade-test.sh`

and it will pass flawlessly. But in the CI workflow it breaks every single time. Any ideas?

**Comment by ZaradarBH** [comment]
> @nghuyenthevinh2000: Please, help! I can run the upgrade test locally on my computer by invoking:
> 
> `make clean upgrade-start-localnet && ./contrib/updates/upgrade-test.sh`
> 
> and it will pass flawlessly. But in the CI workflow it breaks every single time. Any ideas?

It seems to be breaking because this command does not return a block height @ https://github.com/classic-terra/core/blob/58f2c5236c4faf1c5a112c6ba6ab90bc1493e861/contrib/updates/upgrade-test.sh#L41 . This keeps looping for 15 mins until it hits the action timeout @ https://github.com/classic-terra/core/blob/58f2c5236c4faf1c5a112c6ba6ab90bc1493e861/.github/workflows/test.yml#L57

Looking at the error messages it seems like the container its looking for has terminated abnormally, as in the container returned an unexpected exit code. It might be useful to grep the ID of the error msg in the script and run "docker logs {container_id} --details" to figure out what went wrong inside the container. Remember docker has no understanding of what a container does internally, it just fires and forgets, monitors for exit codes and pipe stdout/err to a log, which means you might be able to find the needle in the haystack by looking at the logs of the build agent when it attempts to run the script.

**Pull request merged**